06.bowling_object/game.rb

```diff
@@ -0,0 +1,36 @@
+#!/usr/bin/env ruby
+# frozen_string_literal: true
+
+require_relative 'frame'
+
+class Game
+  def initialize(shot_scores)
+    @frames = [Frame.new(0)]
+
+    shot_scores.gsub('X', '10').split(',').map(&:to_i).each do |shot_score|
+      record_shot(shot_score)
```

Comment:
```suggestion
    shot_scores.gsub('X', '10').split(',').each do |shot_score|
      record_shot(shot_score.to_i)
```

とした方が若干処理効率がいいですね。
元のコードだとmapとeachでそれぞれループが走るので。

Reply:
mapメソッドを利用することで新たな配列が作成されるので、配列2個分のメモリが必要になるのでmapメソッドを利用しないほうが処理効率は良さそうですが、それぞれループが走ることで処理効率が悪くなるというのはどういうことでしょうか？

結局オブジェクトに対して操作する回数は一緒なので、ループで処理するのと1回1回to_iを呼び出すので処理速度は変わらないんじゃないかなーと思いました。

Reply:
ここはshot_scoresの長さがある程度決まっているので、体感できるような速度差はほぼないのですが、仮にshot_scoresの要素が1万件あったとすると、map + eachだとmapで1万回、eachで1万回、合計2万回ループが走ります。
一方、eachだけだとループは1万回で済みます。
基本的にループ処理の回数は少なければ少ないほど処理速度は速くなります。

参考までに以下のようなベンチマークテストを書いてみました。

```ruby
require 'benchmark'

# 数字100文字を文字列を生成
str = Array.new(100) { rand(10) }.join

num_iteration = 10_000

sum_1 = 0
sum_2 = 0
Benchmark.bm 10 do |r|
  r.report "map + each" do
    num_iteration.times do
      str.split.map(&:to_i).each { |x| sum_1 += x }
    end
  end

  r.report "each" do
    num_iteration.times do
      str.split.each { |x| sum_2 += x.to_i }
    end
  end
end
# 念のため合計値が一致することを確認
puts sum_1 == sum_2
```

実行結果

```
                 user     system      total        real
map + each   0.057114   0.000161   0.057275 (  0.057366)
each         0.051818   0.000051   0.051869 (  0.051947)
true
```

コードの意味や結果の見方はこちらの記事を参照。
https://qiita.com/scivola/items/c5b2aeaf7d67a9ef310a

ベンチマークで使ったコードはシンプルなのでそこまで大きな差は出ませんでしたが、処理が複雑になるともう少し大きな差になる可能性があります。

Reply:
オブジェクトに対する処理の回数が重要なのではなく、ループ処理そのものが重い処理なのですね、勉強になりました。
わざわざテストコードまでありがとうございます！